### PR TITLE
add Jaeger tracing

### PR DESCRIPTION
**Pull request by beyang**
Add Jaeger tracing to Zoekt. This replaces the existing `golang.org/x/net/trace.New` calls with calls to `github.com/sourcegraph/zoekt/trace.New`. The latter invokes `net/trace` as before, but also creates an opentracing span.

image

There are two new packages:
* `github.com/sourcegraph/zoekt/trace`: mimics the `net/trace` API, but creates an opentracing span under the hood.
* `github.com/sourcegraph/zoekt/trace/ot`: a wrapper around the opentracing API that enables selective tracing toggled by a context item.

These two packages are very similar to `github.com/sourcegraph/sourcegraph/internal/trace` and `github.com/sourcegraph/sourcegraph/internal/ot`. For now, it seems like they are different enough that it isn't worth the effort of merging them (which would mean breaking these packages out into a separate repository). If we find that we are frequently making changes to one and having to reflect them over to the other, we can revisit down the road.

This PR has a sister PR in the main Sourcegraph repo (https://github.com/sourcegraph/sourcegraph/pull/12548), which should be merged immediately after this is merged and it is updated with the latest dep version.

~Note that this does not yet propagate span IDs across the RPC boundary between `frontend` and `zoekt`, so the `zoekt` spans are "dangling".~ SpanContext is now propagated across the RPC boundary, so the zoekt spans are properly "attached" in the trace.

Questions:
* @keegancsmith this updates the RPC options struct. Will this break RPC calls between `frontend` and `zoekt-webserver` if this is not deployed atomically with the corresponding update to `frontend`?

Merging this will help with the effort to add more repos to the Sourcegraph.com global index.

## Reviewer guide: :point_right: [Start here](https://github.com/sourcegraph/zoekt/pull/55/files#r463240093)

### CODE REVIEW

**Comment by beyang** [comment]
From @keegancsmith:

> I'm not exactly sure on the value of tracing each shard access (yet). Tracing in the server side sounds good. But it won't contribute that much over what we are currently tracing from the client side (in frontend). This is because Zoekt returns stats about what it did in its response. I also recently improved the logging of that a lot. For example see the logs in this trace: https://sourcegraph.com/-/debug/jaeger/trace/54aa2901efff10e3?uiFind=4a9e3ea58ec2c84b
> 
> What are your thoughts on what is currently done vs adding tracing into zoekt + slightly complicating the binary. If we do want to add jaeger support in, I think we should consider building the zoekt binary in sourcegraph so we get access to internal/zoekt and conf pkg.

**Comment by beyang** [comment]
> What are your thoughts on what is currently done vs adding tracing into zoekt + slightly complicating the binary.

As someone who is less familiar with Zoekt internals, having the Jaeger tracing helps me understand what's going on vs. I'm not exactly sure how to interpret the list of stats that's logged in the frontend-side trace. It also shows if one particular index file took much longer to read (as in the case of the screenshot in the PR description), whereas I'm not sure how to reproduce this information from the aggregate stats in what's currently logged.

> If we do want to add jaeger support in, I think we should consider building the zoekt binary in sourcegraph so we get access to internal/zoekt and conf pkg.

That seems like a good idea. Would that just involve moving `cmd/zoekt-webserver` into the Sourcegraph main repo? It would be nice to have zoekt have access to the conf pkg... but I think the overarching concern I have at the moment is increasing the size of the sourcegraph.com global search index—and determining whether Zoekt is the right solution for scaling up to 1M repositories. If it's not, then the work we do moving it into the Sourcegraph main repo would be a bit of a waste. Therefore, I'd like to defer moving it into the main repo for the time being.

TL;DR: this would help me with understanding Zoekt behavior so we can continue increasing the size of the global repo search index on sourcegraph.com. Happy to rip all this out if it later turns out not to be useful.

File: cmd/zoekt-webserver/main.go [hunk 1]
```diff
@@ -124,8 +131,34 @@ func main() {
 	templateDir := flag.String("template_dir", "", "set directory from which to load custom .html.tpl template files")
 	dumpTemplates := flag.Bool("dump_templates", false, "dump templates into --template_dir and exit.")
 	version := flag.Bool("version", false, "Print version number")
+	useJaeger := flag.Bool("jaeger", false, "Collect and send traces to Jaeger. Jaeger configuration set via environment variables. Only traces for RPC requests that explicitly opt into traces will be collected.")
```
Comment (beyang) [inline cmd/zoekt-webserver/main.go]: Command-line flag enables Jaeger by setting the GlobalTracer in the `github.com/opentracing/opentracing-go` package.

[Back](https://github.com/sourcegraph/zoekt/pull/55/files#r463239342)
[Next](https://github.com/sourcegraph/zoekt/pull/55/files#r463249224)

File: trace/ot/ot.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
+// Package ot wraps github.com/opentracing/opentracing-go and
```
Comment (beyang) [inline trace/ot/ot.go]: And here is the implementation of that `ot` package, which wraps `github.com/opentracing/opentracing-go`. It is a partial copy of the `ot` package in the Sourcegraph main repo.

[Back](https://github.com/sourcegraph/zoekt/pull/55/files#r463239415) [Next](https://github.com/sourcegraph/zoekt/pull/55/files#r463239081)

File: trace/trace.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
+// Package trace provides a tracing API that in turn invokes both the `golang.org/x/net/trace` API
```
Comment (beyang) [inline trace/trace.go:1]: And here is the `github.com/sourcegraph/zoekt/trace` package, mostly copied from `trace` package in the Sourcegraph main repo. It mimics the API of `golang.org/x/net/trace`.

Underneath the hood, it makes use of `github.com/sourcegraph/zoekt/trace/ot`, which is partially copied from `github.com/sourcegraph/sourcegraph/internal/trace/ot`. This package is a wrapper around the `github.com/opentracing/opentracing-go` package that implements selective tracing (so traces are sent only for requests that opt into tracing via a context item).

[Back](https://github.com/sourcegraph/zoekt/pull/55/files#r463244129) [Next](https://github.com/sourcegraph/zoekt/pull/55/files#r463239342)

File: web/server.go [hunk 4]
```diff
@@ -170,7 +170,7 @@ func NewMux(s *Server) (*http.ServeMux, error) {
 		mux.HandleFunc("/print", s.servePrint)
 	}
 	if s.RPC {
-		mux.Handle(rpc.DefaultRPCPath, rpc.Server(s.Searcher)) // /rpc
+		mux.Handle(rpc.DefaultRPCPath, rpc.Server(traceAwareSearcher{s.Searcher})) // /rpc
```
Comment (beyang) [inline web/server.go:173]: :point_right: Reviewer, start here.

This is where we intercept the RPC handler to set up things for tracing.

[Next](https://github.com/sourcegraph/zoekt/pull/55/files#r463242530)

File: web/trace.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/google/zoekt/query"
+	"github.com/google/zoekt/trace/ot"
+	"github.com/opentracing/opentracing-go"
+	"golang.org/x/net/context"
+)
+
+// traceAwareSearcher wraps a zoekt.Searcher instance so that the tracing context item is set in the
+// context. This context item toggles on trace collection via the
+// github.com/sourcegraph/zoekt/trace/ot package.
+type traceAwareSearcher struct {
```
Comment (beyang) [inline web/trace.go:16]: This struct wraps the underlying `zoekt.Searcher` instance. The only handler that does anything is `Search` (the other methods are just pass-throughs). `Search` adds 2 items to the context: (1) whether we should trace at all, according to `frontend` and (2) the SpanContext if we are tracing.

[Back](https://github.com/sourcegraph/zoekt/pull/55/files#r463240093) [Next](https://github.com/sourcegraph/zoekt/pull/55/files#r463244129)

File: eval.go [hunk 6]
```diff
@@ -104,7 +103,7 @@ func (d *indexData) Search(ctx context.Context, q query.Q, opts *SearchOptions)
 	default:
 	}
 
-	tr := trace.New("indexData.Search", d.file.Name())
+	tr, ctx := trace.New(ctx, "indexData.Search", d.file.Name())
```
Comment (beyang) [inline eval.go]: The `github.com/google/zoekt/trace` package replaces `golang.org/x/net/trace` and presents the same API. This is the same pattern as in the main repository ([example](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/search.go?subtree=true#L74:17)).
[Back](https://github.com/sourcegraph/zoekt/pull/55/files#r463242530) [Next](https://github.com/sourcegraph/zoekt/pull/55/files#r463239415)

File: api.go [hunk 7]
```diff
@@ -316,6 +316,13 @@ type SearchOptions struct {
 	// Trim the number of results after collating and sorting the
 	// results
 	MaxDocDisplayCount int
+
```
Comment (beyang) [inline api.go:319]: The struct fields in the `SearchOptions` struct, which are passed through the RPC boundary. To see how they are used by the RPC client (`frontend`), go to the sister PR: [Next](https://github.com/sourcegraph/sourcegraph/pull/12548/files#diff-39413ee2a9c32387903e199473779d74R216-R217)

[Back](https://github.com/sourcegraph/zoekt/pull/55/files#r463239081)

**Comment by keegancsmith** [comment]
Really sorry I did't get around to reviewing this properly today. I promise to tackle it Monday.

Without actually reading the code these are my thoughts:

1. The RPC connection is stateful. You can communicate jaeger configuration from the frontend that is attached to the connection. That should quite nicely work I think? Unless jaeger is better with a global tracer.
2. I saw mention of communicating sampling strategy. Is that important? Shouldn't it be as simple as if an RPC has a span attached, use that span.
3. "Will this break RPC calls between frontend and zoekt-webserver". No, gob is backwards compatible for "missing" fields. https://golang.org/pkg/encoding/gob/#hdr-Types_and_Values
4. My main concern is introducing a performance penalty. For example when searching a shard that has no trigram matches zoekt does _very_ little work. I would want to ensure jaeger doesn't do much if no one ever looks at the trace. For example `net/trace` is really good for that. The `lazyprintf` only ever evaluates a printf if someone actually looks at it.
5. Tracing per shard is not that useful for understanding performance problems at scale. To be honest I can imagine this amount of tracing data will just overload the trace and make it near impossible to understand (or even load) any real trace. I'd be much happier if we just traced at the shardedSearcher level rather than at the index level.
6. The packages you add should live under `internal`, or we should consider adding a new repo with our nice `internal/trace` pkg :)

**Review by uwedeportivo — COMMENTED** [review]
LGTM and i love the navigation

i'll let keegan accept

File: cmd/zoekt-webserver/main.go (hunk 1 above)
Comment (keegancsmith) [reply cmd/zoekt-webserver/main.go]: Do we ever want this not to be true in our context? I would avoid adding a flag and just always enable if that works.

File: cmd/zoekt-webserver/main.go [hunk 8]
```diff
@@ -124,8 +131,34 @@ func main() {
...
+		}
+		tracer, _, err := cfg.NewTracer(
+			jaegercfg.Logger(jaegerlog.StdLogger),
+			jaegercfg.Metrics(jaegermetrics.NullFactory),
+		)
+		if err != nil {
+			log.Printf("could not initialize jaeger tracer, error: %v", err.Error())
+		}
+		opentracing.SetGlobalTracer(tracer)
+	}
```
Comment (keegancsmith) [inline cmd/zoekt-webserver/main.go]: nit: this is a fair number of lines, probably deserves its own function since it doesn't actually mutate/return anything in main.

File: trace/ot/ot.go (hunk 2 above)
Comment (keegancsmith) [reply trace/ot/ot.go]: I don't really understand why we need this. Surely we should trace iff we get a spancontext. Otherwise we don't. We don't need the same decision logic as what processes incoming requests in the frontend.

IE at the RPC layer we add in the spancontext to the `context.Context` if it is specified. Otherwise when creating a new trace we skip the OT part if there is not span part of the passed in ctx. Does that make sense? Should reduce amount of code for this PR by quite a bit.

File: trace/trace.go (hunk 3 above)
Comment (keegancsmith) [reply trace/trace.go:1]: I'd be happy putting this into a new repo `github.com/sourcegraph/trace`. For example since you created this PR I updated the trace pkg in sourcegraph to be a bit better. I then improved how we log around zoekt results which would be useful to have here: https://github.com/sourcegraph/sourcegraph/pull/12554

File: web/trace.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+			span, newCtx := ot.StartSpanFromContext(ctx, "zoekt.traceAwareSearcher.Search", opentracing.ChildOf(spanContext))
+			defer span.Finish()
+			ctx = newCtx
+		}
+	}
+	return s.Searcher.Search(ctx, q, opts)
+}
+
+func (s traceAwareSearcher) List(ctx context.Context, q query.Q) (*zoekt.RepoList, error) {
+	return s.Searcher.List(ctx, q)
```
Comment (keegancsmith) [inline web/trace.go:37]: this is also useful to trace.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
The main blocker to landing this is `net/trace` is fine with per shard logging, but `ot` is not. If we turn this on we won't be able to load a trace for a search on sourcegraph.com since it will search over 100k+ shards => 100k+ spans of noise.

I'm not exactly sure on the solution to this. I think it is likely making it possible at the client call site for `trace` to opt out of `ot`.

File: cmd/zoekt-webserver/main.go (hunk 1 above)
Comment (beyang) [reply cmd/zoekt-webserver/main.go]: I think it would be nice to enable/disable Jaeger without shipping a new binary, and I think the flag adds a minor amount of complexity (removing it would eliminate ~2 lines of code).

**Comment by beyang** [comment]
@keegancsmith I meant to do more work on this PR today, but got sidetracked, so just wanted to post what my plans are, since I only partially responded to the review comments:
* Will remove the tracing for index shards, since I agree that has a chance of causing a decent perf hit.
* I am hesitant to factor out the `trace` and `ot` packages into a separate repository, because it feels a bit premature to generalize here. The tracing needs of the Sourcegraph main repo differ a bit from what we want in Zoekt. For instance, the behavior of "only record spans if the context already contains a parent span" is not needed in the main repo's `trace` package, which is toggled by the tracing mode package global variable (`all`, `none`, `selective`). I think I'd rather have a package specific to this repository that satisfies *its* need for now. The existing `trace` and `ot` packages in this repository can be simplified to meet that requirement (e.g., no need for the tracing mode package global variable copied over from the main repo, as you point out), which I will do. In other words, I will simplify these packages in this repository, but I think it is better *not* to factor out the tracing packages into a separate dependency for now.
* I think it's better to keep the command-line flag to turn on/off Jaeger. It doesn't add *that* much complexity and it may come in handy in case we need to turn Jaeger off for whatever reason on sourcegraph.com or on a customer instance.
* The other issues you pointed out I think are minor and I'll address/fix them.

^ Lemme know if you have any objections to the above plan of action.

**Comment by keegancsmith** [comment]
> * Will remove the tracing for index shards, since I agree that has a chance of causing a decent perf hit.

Great.

> * I am hesitant to factor out the `trace` and `ot` packages into a separate repository, because it feels a bit premature to generalize here. The tracing needs of the Sourcegraph main repo differ a bit from what we want in Zoekt. For instance, the behavior of "only record spans if the context already contains a parent span" is not needed in the main repo's `trace` package, which is toggled by the tracing mode package global variable (`all`, `none`, `selective`). I think I'd rather have a package specific to this repository that satisfies _its_ need for now. The existing `trace` and `ot` packages in this repository can be simplified to meet that requirement (e.g., no need for the tracing mode package global variable copied over from the main repo, as you point out), which I will do. In other words, I will simplify these packages in this repository, but I think it is better _not_ to factor out the tracing packages into a separate dependency for now.

Sounds good!

> * I think it's better to keep the command-line flag to turn on/off Jaeger. It doesn't add _that_ much complexity and it may come in handy in case we need to turn Jaeger off for whatever reason on sourcegraph.com or on a customer instance.

Adjusting the command line for docker containers is a pain in the arse to be honest. Probably better to make it an environment variable.

**Comment by beyang** [comment]
@keegancsmith PTAL

* I moved the `ot` package into `trace`. After simplifying, there were only a few symbols remaining, so it made sense to merge.
* I would like to punt adding tracing to `traceAwareSearcher.List` to a subsequent PR to be done immediately after this is merged (because I'd like to get this deployed in the next day or so).
* Removed opentracing for index shards.
* Removed the Jaeger flag, now using an environment variable (`JAEGER_DISABLED`, which is interpreted as default true if unset).

If remaining comments are minor, I'd prefer to address them in a follow-up PR if that's alright.

File: cmd/zoekt-webserver/main.go [hunk 10]
```diff
@@ -314,3 +324,40 @@ func (s *loggedSearcher) Search(ctx context.Context, q query.Q, opts *zoekt.Sear
...
+	isJaegerDisabled, err := strconv.ParseBool(jaegerDisabled)
+	if err != nil {
+		log.Printf("EROR: failed to parse JAEGER_DISABLED: %s", err)
+		return
+	}
+	if isJaegerDisabled {
+		return
+	}
+	log.Printf("connecting to Jaeger")
```
Comment (keegancsmith) [inline cmd/zoekt-webserver/main.go]: nit: this looks like a low value log message

**Review by keegancsmith — APPROVED** [review]
